vm/src/compiler.rs

```diff
@@ -719,10 +719,21 @@ impl<'a> Compiler<'a> {
                         Pattern::Record { .. } => {
                             start_jumps.push(function.function.instructions.len());
                         }
-                        _ => {
+                        Pattern::Ident(_) => {
                             start_jumps.push(function.function.instructions.len());
                             function.emit(Jump(0));
                         }
+                        Pattern::Literal(ref l) => {
+                            let instruction = match *l {
+                                ast::Literal::Byte(_) => ByteEQ,
+                                ast::Literal::Int(_) | ast::Literal::Char(_) => IntEQ,
+                                ast::Literal::Float(_) => FloatEQ,
+                                ast::Literal::String(_) => unimplemented!(),
+                            };
+                            function.emit(instruction);
+                            start_jumps.push(function.function.instructions.len());
+                            function.emit(CJump(0));
```

Comment:
What you are missing here is to push the arguments to the instruction to the top of the stack.

The value that we are inspecting (the value between `match ... with`) should be at the top of the stack here so emit a `Push(function.stack_size() - 1)` instruction for that.

For the literal itself it is easy for bytes,chars and floats, just emit `PushInt/PushFloat`. For strings you can instead call `emit_string` and it will make sure to add the string constant to the environment and emit the `PushString` instruction.

After those values are pushed you can emit the comparison.

```
// scrutinee is what is between match scrutinee with
[..., scrutinee]
Push(stack_size - 1)
[..., scrutinee, scrutinee]
PushInt(123)
[..., scrutinee, scrutinee, 123]
IntEQ
[..., scrutinee, bool]
CJump
[..., scrutinee]
// Note that we couldn't skip the `Push` instruction since wee need scrutinee on the top of the stack after CJump executes since we may have another alternative after that compares against a different value
```